dist/tools/bmp/bmp.py

```diff
@@ -0,0 +1,274 @@
+        total_size = int(m.group(5)) if m.group(5) else None
+        return section_name, section_sent, section_size, total_sent, total_size
+
+
+def download_to_flash(gdbmi):
+    # download to flash
+    res = gdbmi.write('-target-download', timeout_sec=TIMEOUT)
+    first = True  # whether this is the first status message
+    current_sec = None  # name of current section
+    pbar = ProgressBar()
+    while True:
+        for msg in res:
+            if msg['type'] == 'result':
+                assert msg['message'] == 'done', "download failed: %s" % str(msg)
+                if pbar.start_time:
+                    pbar.finish()
+                print("downloading finished")
+                return
+            elif msg['type'] == 'output':
+                section_name, section_sent, section_size, total_sent, total_size = parse_download_msg(msg)
+                if section_name:
+                    if first:
+                        first = False
+                        print("downloading... total size: %s"
+                              % humanize.naturalsize(total_size, gnu=True))
+                    if section_name != current_sec:
+                        if pbar.start_time:
+                            pbar.finish()
+                        current_sec = section_name
+                        print("downloading section [%s] (%s)" % (
+                            section_name, humanize.naturalsize(section_size, gnu=True)))
+                        pbar = ProgressBar(widgets=[Percentage(), Bar()], maxval=section_size).start()
+                    if section_sent:
+                        pbar.update(section_sent)
+        res = gdbmi.get_gdb_response(timeout_sec=TIMEOUT)
+
+
+def check_flash(gdbmi):
+    res = gdbmi.write('compare-sections', timeout_sec=TIMEOUT)
+    while True:
+        for msg in res:
+            if msg['type'] == 'result':
+                assert msg['message'] == 'done', "checking failed: %s" % str(msg)
+                print("checking successful")
+                return
+            elif msg['type'] == 'console':
+                assert 'matched' in msg['payload'] and 'MIS-MATCHED' not in msg['payload'], \
+                    "checking failed: %s" % str(msg)
+        res = gdbmi.get_gdb_response(timeout_sec=TIMEOUT)
+
+
+if __name__ == '__main__':
+    args = parser.parse_args()
+    assert not (args.swd and args.jtag), "you may only choose one protocol"
+    assert not (args.serial and args.port), "you may only specify the probe by port or by serial"
+    g, u = detect_probes()
+    assert len(g) > 0, "no Black Magic Probes found 😔"
+
+    # terminal mode, opens TTY program
+    if args.action == 'term':
+        port = u[0].device
+        if args.port:
+            port = args.port
+        elif args.serial:
+            port = search_serial(args.serial, u)
+            assert port, "no BMP with this serial found"
+        os.system(args.term_cmd % port)
+        sys.exit(0)
```

Comment:
How about having the individual actions as separate functions, e.g. something like `term_mode()`?

Reply:
this part became `term_mode(uart_ports)`